Early return in the slow path of `put` if key already exists and `no_replacement` is true

domenicquirl opened the pull request:
Early return with an indication that the requested key already existed and a reference back to the value trying to be inserted.

Fixes #90. 

I also took the liberty to add a new `issues.rs` test file to contain reproducible bugs that should now be fixed, so we will know should they appear again.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/jonhoo/flurry/91)
<!-- Reviewable:end -->

domenicquirl commented:
Interesting. The failure on the warning screening seems to be because `#[deny(intra_doc_link_resolution_failure)]` was renamed?

jonhoo commented:
Ah, good catch! And :+1: to adding `issues.rs`. Maybe call it `regressions` though?

And yes, the lint has been renamed to `broken_intra_doc_links` (https://github.com/rust-lang/rust/pull/74926).

domenicquirl commented:
Added the suggested name change and also updated the lint. Interestingly, on my machine I get the warning for the lint the other way round now :D 
Hopefully it's good now on CI.

jonhoo commented:
Ah, it's probably just because you have a slightly older nightly locally which doesn't have that latest rename PR merged!

domenicquirl commented:
If it's only changed on nightly than that'll be it, I'm on stable for most things that don't rely on unstable features 👍

domenicquirl commented:
Everything but Miri ran I think, and Miri timed out last time if I remember correctly - you fine with merging?

jonhoo commented:
Hmm, miri timing out isn't great. My guess is that it's because this new regression test runs for so many iterations. Iif the test truly _requires_ this many iterations to reproduce the issue, then just use a different number of iterations for miri. You can use `cfg!(miri)` for this. Otherwise, just reduce the number of iterations and miri should be fine.

domenicquirl commented:
Oh, you think? I thought some other stuff like the stress test were pretty large as well, but seems we already have some miri-dependent tests too. Updating again.

This keeps the original number for non-miri tests, since it runs pretty quickly and the intention is to (hopefully) exercise both the regular and the tree bin path, for which we need some collisions.

jonhoo commented:
I believe the stress tests are already `#[cfg(not(miri))]`, but could be wrong?

domenicquirl commented:
They are ignored completely it seems. Other tests have smaller ranges and such. In any case, seems to have been it!
